ecosystem/indexer-grpc/indexer-grpc-data-service/src/service.rs

```diff
@@ -416,15 +190,401 @@ impl RawData for RawDataServerWrapper {
+            async move {
+                get_data_in_thread(
+                    current_version,
+                    chain_id,
+                    &mut cache_operator,
+                    file_store_operator,
+                    request_metadata.clone(),
+                    cache_storage_format,
+                )
+                .await
+            }
+        });
+        threads.push(thread);
+        current_version += TRANSACTIONS_PER_STORAGE_BLOCK;
+    }
+
+    let mut transactions: Vec<Vec<Transaction>> = vec![];
+    join_all(threads)
+        .await
+        .into_iter()
+        .for_each(|result| match result {
+            Ok(DataFetchSubThreadResult::Success(txns)) => {
+                transactions.push(txns);
+            },
+            Ok(DataFetchSubThreadResult::NoResults) => {},
+            Err(e) => {
+                error!(
+                    error = e.to_string(),
+                    "[Data Service] Failed to get data from cache and file store."
+                );
+                panic!("Failed to get data from cache and file store.");
+            },
+            Ok(_) => unreachable!("Fetching from a single thread will never return a batch"),
+        });
+
+    if transactions.is_empty() {
+        DataFetchSubThreadResult::NoResults
+    } else {
+        DataFetchSubThreadResult::BatchSuccess(transactions)
+    }
+}
+
+async fn get_data_in_thread(
+    start_version: u64,
+    chain_id: u64,
+    cache_operator: &mut CacheOperator<redis::aio::ConnectionManager>,
+    file_store_operator: Arc<Box<dyn FileStoreOperator>>,
+    request_metadata: Arc<IndexerGrpcRequestMetadata>,
+    cache_storage_format: StorageFormat,
+) -> DataFetchSubThreadResult {
+    let current_batch_start_time = std::time::Instant::now();
+    let transaction_data = match data_fetch(
+        start_version,
+        cache_operator,
+        file_store_operator,
+        request_metadata.clone(),
+        cache_storage_format,
+    )
+    .await
+    {
+        Ok(TransactionsDataStatus::Success(transactions)) => transactions,
+        Ok(TransactionsDataStatus::AheadOfCache) => {
+            info!(
+                start_version = start_version,
+                request_name = request_metadata.processor_name.as_str(),
+                request_email = request_metadata.request_email.as_str(),
+                request_api_key_name = request_metadata.request_api_key_name.as_str(),
+                processor_name = request_metadata.processor_name.as_str(),
+                connection_id = request_metadata.request_connection_id.as_str(),
+                request_user_classification = request_metadata.request_user_classification.as_str(),
+                duration_in_secs = current_batch_start_time.elapsed().as_secs_f64(),
+                service_type = SERVICE_TYPE,
+                "[Data Service] Requested data is ahead of cache. Sleeping for {} ms.",
+                AHEAD_OF_CACHE_RETRY_SLEEP_DURATION_MS,
+            );
```

Comment:
I had a PR to remove this, can you remove it here?

Reply:
I'll rebase before merge